                                                Asssignement: Neural Networks 

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score, recall_score, f1_score, confusion_matrix



In [3]:
df = pd.read_csv("C:\\Ankit_Singh\\Data Science\\Sample Data\\Alphabets_data.csv")
df.head()


letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


In [5]:
# Normalizing the features
features = df.drop(columns=['letter'])  
scaler = StandardScaler()
features_scaled = scaler.fit_transform(features)  


In [6]:
# Encode labels (letters) into numeric values

labels = df['letter'].factorize()[0]


In [7]:
# Now let's segregate the dataset into two parts
X_train, X_test, y_train, y_test = train_test_split(features_scaled, labels, test_size=0.2, random_state=42)


In [8]:
pip install tensorflow

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [9]:
pip install keras

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [10]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam


In [11]:
# Now let's try to build a model
model = Sequential([
    Dense(128, input_dim=X_train.shape[1], activation='relu'),
    Dense(64, activation='relu'),
    Dense(len(np.unique(labels)), activation='softmax')  
])


c:\Users\akash\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [13]:
model_compile=model.compile(optimizer=Adam(learning_rate=0.001), loss='sparse_categorical_crossentropy', metrics=['accuracy'])


In [14]:
Train_model = model.fit(X_train, y_train, epochs=50, batch_size=32, validation_split=0.2)


Epoch 1/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.4301 - loss: 2.1725 - val_accuracy: 0.7444 - val_loss: 0.8941
Epoch 2/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7829 - loss: 0.7658 - val_accuracy: 0.8109 - val_loss: 0.6413
Epoch 3/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8401 - loss: 0.5665 - val_accuracy: 0.8522 - val_loss: 0.5178
Epoch 4/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8659 - loss: 0.4608 - val_accuracy: 0.8753 - val_loss: 0.4378
Epoch 5/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8840 - loss: 0.3994 - val_accuracy: 0.8838 - val_loss: 0.3891
Epoch 6/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9009 - loss: 0.3352 - val_accuracy: 0.9000 - val_loss: 0.3416
Epoch 7/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9165 - loss: 0.2860 - val_accuracy: 0.9122 - val_loss: 0.3080
Epoch 8/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9273 - loss: 0.2459 - val_accuracy: 0.

In [15]:
import keras_tuner as kt


In [16]:
def build_model(hp):
    model = Sequential()
    model.add(Dense(units=hp.Int('units', min_value=64, max_value=256, step=64),
                    activation='relu', input_dim=X_train.shape[1]))
    model.add(Dense(units=hp.Int('units', min_value=64, max_value=256, step=64), activation='relu'))
    model.add(Dense(len(np.unique(labels)), activation='softmax'))
    
    model.compile(optimizer=Adam(learning_rate=hp.Float('learning_rate', min_value=1e-5, max_value=1e-2, sampling='log')),
                  loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    
    return model



In [17]:
tuner = kt.Hyperband(build_model, objective='val_accuracy', max_epochs=10, factor=3, directory='my_dir', project_name='alphabet_tuning')


Reloading Tuner from my_dir\alphabet_tuning\tuner0.json


In [18]:
tuner.search(X_train, y_train, epochs=10, validation_data=(X_test, y_test))


In [19]:
# Now let's find out what is the best Hyperparameter
best_hp = tuner.get_best_hyperparameters()[0]
print("Best Hyperparameters:", best_hp.values)


Best Hyperparameters: {'units': 256, 'learning_rate': 0.0035815894446913792, 'tuner/epochs': 10, 'tuner/initial_epoch': 4, 'tuner/bracket': 1, 'tuner/round': 1, 'tuner/trial_id': '0022'}


In [22]:
best_model = tuner.get_best_models(num_models=1)[0]

test_loss, test_accuracy = best_model.evaluate(X_test, y_test)

print(f'Test Loss: {test_loss:.4f}')
print(f'Test Accuracy: {test_accuracy:.4f}')



  1/125 ━━━━━━━━━━━━━━━━━━━━ 17s 137ms/step - accuracy: 1.0000 - loss: 0.0459

c:\Users\akash\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\akash\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\saving\saving_lib.py:757: UserWarning: Skipping variable loading for optimizer 'adam', because it has 2 variables whereas the saved optimizer has 14 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9637 - loss: 0.1181
Test Loss: 0.1217
Test Accuracy: 0.9622


In [24]:
# Since we train our model let's try to make some prediction
y_pred = np.argmax(best_model.predict(X_test), axis=-1)


125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 669us/step


In [25]:
# now to check if our model is really working well let's compute Precision,Recall,F1-score
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
f1 = f1_score(y_test, y_pred, average='weighted')


In [26]:
# Display confusion matrix and classification metrics
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)

print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1 Score: {f1:.4f}")


Confusion Matrix:
[[160   0   0   0   0   0   1   0   0   0   0   0   0   1   0   0   0   0
    0   0   0   0   0   0   0   1]
 [  0 143   1   0   0   1   0   0   0   0   1   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0]
 [  0   0 153   2   1   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0]
 [  0   0   0 145   0   0   0   0   0   0   0   1   4   1   0   0   0   0
    0   0   0   0   0   0   0   0]
 [  0   0   3   0 149   0   2   0   0   0   0   0   0   0   3   0   0   1
    0   2   0   0   0   0   0   0]
 [  0   0   0   0   0 170   1   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0]
 [  2   0   0   0   0   0 149   0   0   0   0   0   0   1   0   0   0   0
    0   0   1   0   0   0   0   0]
 [  0   0   0   0   0   0   0 148   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   1   0   0]
 [  0   5   0   0   0   0   0   0 142   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   2]
 

Observations:

Dataset Overview:

The dataset contains 20,000 samples with 16 numerical features and 1 categorical target (letter).

All feature columns are numeric and non-null, making the dataset clean and ready for modeling.

Preprocessing:

Features were standardized using StandardScaler.

The categorical target (letter labels) was encoded into numeric values.

Data was split into training and test sets using an 80-20 split.

Model Building and Training:

A Neural Network was built using TensorFlow/Keras with an architecture of 128 → 64 → softmax output layer.

Initial model was trained for 50 epochs with a validation split of 0.2.

Hyperparameter tuning was performed using Keras Tuner (Hyperband strategy) optimizing for validation accuracy.

The best model used:

Units: 256 per layer

Learning rate: ~0.00358

Evaluation:

The best model achieved:

Test Accuracy: 96.22%

Test Loss: 0.1217

Predictions were made on the test set.

Performance metrics:

Precision: 0.9631

Recall: 0.9623

F1 Score: 0.9621

The confusion matrix revealed very low misclassification across most letters, showing high reliability.

Conclusion Statement:
The neural network model demonstrated strong performance in classifying handwritten alphabet characters with over 96% accuracy. The precision, recall, and F1 scores are all above 96%, indicating a well-balanced model with high classification effectiveness. Hyperparameter tuning significantly improved model accuracy, showing the importance of optimization in deep learning workflows. Overall, the model generalizes well and can be reliably used for character recognition tasks on similar datasets.

